### PR TITLE
List tags and copy images between registries

### PR DESCRIPTION
**Pull request by lenardchristopher**
This PR implements pulling, tagging and pushing images. We are using an awesome repo from Google called [go-containerregistry](https://github.com/google/go-containerregistry).

In the repo, is a tool called [Crane](https://github.com/google/go-containerregistry/tree/main/cmd/crane), which wraps all the registry API logic in an intuitive way. With it, the image promotion logic is wrapped up very succinctly. Check out `pkg/processor.go` for the image promotion logic.

This was tested successfully with two scenarios
- busybox between two DockerHub repos. (2.71 MB, <2 seconds)
- a large image between two azure container registries  (4.69 GB, took ~7 minutes)

### CODE REVIEW

File: .devcontainer/Dockerfile [hunk 1]
```diff
@@ -46,4 +46,9 @@ RUN curl -LO https://github.com/operator-framework/operator-sdk/releases/downloa
 ENV KUBECTL_VERSION=1.19.11
 RUN curl -LO https://storage.googleapis.com/kubernetes-release/release/v${KUBECTL_VERSION}/bin/linux/amd64/kubectl \
   && chmod +x ./kubectl \
-  && sudo mv ./kubectl /usr/local/bin/kubectl
\ No newline at end of file
+  && sudo mv ./kubectl /usr/local/bin/kubectl
+  
+# Install Kind
+RUN curl -Lo ./kind https://kind.sigs.k8s.io/dl/v0.11.0/kind-linux-amd64 \
```
Comment (lenardchristopher) [inline .devcontainer/Dockerfile:52]: Add kind for easy local functional testing

File: .vscode/launch.json [hunk 2]
```diff
@@ -0,0 +1,16 @@
+{
+    // Use IntelliSense to learn about possible attributes.
+    // Hover to view descriptions of existing attributes.
+    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
+    "version": "0.2.0",
+    "configurations": [
```
Comment (lenardchristopher) [inline .vscode/launch.json:6]: This a vscode launch config for easy local native debugging

File: controllers/containermanagementobject_controller.go [hunk 3]
```diff
@@ -112,9 +114,17 @@ func (r *ContainerManagementObjectReconciler) Reconcile(ctx context.Context, req
...
 		}
+
+		registryCreds = append(registryCreds, *config)
 	}
 
-	// TODO: Call image promotion processor logic
+	// Run image promotion processor
+	client := processor.NewRegistryClient()
+
+	err = processor.Run(client, instance.Spec.Images, registryCreds)
```
Comment (lenardchristopher) [inline controllers/containermanagementobject_controller.go:124]: Dependency injection. We are passing in the implementation of the registry client. This makes it possible to mock the implementation for unit testing.

File: pkg/processor/processor_test.go [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+				&remote.Descriptor{},
+			},
+			wantErr: false,
+		},
+	}
+
+	for _, tt := range tests {
+		mockRegistryClient := &MockRegistryClient{}
+
+		mockRegistryClient.On("listTags", tt.args.images[0].TargetRepository, tt.args.registries[0]).Return(tt.mockReturns[0], nil)
```
Comment (lenardchristopher) [inline pkg/processor/processor_test.go:129]: Check out this cool way I learned how to mock. Mockery helps generate all the code and setting up the return values is quick and readable.

**Review by justinbanks — APPROVED** [review]

**Review by justinbanks — APPROVED** [review]

**Review by coreywagehoft — APPROVED** [review]

**Pull request merged**